src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java

```diff
@@ -150,13 +151,56 @@ public boolean generateTestCasesFromFeedbacks(List<Feedback> feedbacks, Programm
      * @return Result with updated feedbacks, score and result string.
      */
     public Result updateResultFromTestCases(Result result, ProgrammingExercise exercise, boolean isStudentParticipation) {
-        boolean shouldTestsWithAfterDueDateFlagBeRemoved = isStudentParticipation && exercise.getBuildAndTestStudentSubmissionsAfterDueDate() != null
-                && ZonedDateTime.now().isBefore(exercise.getBuildAndTestStudentSubmissionsAfterDueDate());
         Set<ProgrammingExerciseTestCase> testCases = findActiveByExerciseId(exercise.getId());
+        Set<ProgrammingExerciseTestCase> testCasesForCurrentDate = testCases;
+        // We don't filter the test cases for the solution/template participation's results as they are used as indicators for the instructor!
+        if (isStudentParticipation) {
+            testCasesForCurrentDate = filterTestCasesForCurrentDate(exercise, testCases);
+        }
+        return updateResultFromTestCases(testCases, testCasesForCurrentDate, result);
+    }
+
+    /**
+     * Updates <b>all</b> latest automatic results of the given exercise with the information of the exercises test cases. This update includes:
+     * - Checking which test cases were not executed as this is not part of the bamboo build (not all test cases are executed in an exercise with sequential test runs)
+     * - Checking the due date and the afterDueDate flag
+     * - Recalculating the score based based on the successful test cases weight vs the total weight of all test cases.
+     *
+     * If there are no test cases stored in the database for the given exercise (i.e. we have a legacy exercise) or the weight has not been changed, then the result will not change
+     *
+     * @param exercise the exercise whose results should be updated
+     * @return the results of the exercise that have been updated
+     */
+    public List<Result> updateAllResultsFromTestCases(ProgrammingExercise exercise) {
+        Set<ProgrammingExerciseTestCase> testCases = findActiveByExerciseId(exercise.getId());
+        Set<ProgrammingExerciseTestCase> testCasesForCurrentDate = filterTestCasesForCurrentDate(exercise, testCases);
+        // We only update the latest automatic results here, later manual assessments are not affected
+        List<StudentParticipation> participations = participationService.findByExerciseIdWithLatestAutomaticResult(exercise.getId());
```

Comment:
should we update the `templateParticipation` and the `solutionParticipation` as well?
While they should have 0% and 100% anyway, this is not always the case, so I guess this would make sense.
Especially when there is a bonus, the solution result might change (e.g. from 100% to 125%)

Reply:
Template and solution are now updated as well.